src/third/redux-persist/getStoredState.js

```diff
@@ -18,27 +18,30 @@ export default function getStoredState (config, onComplete) {
   let restoredState = {}
   let completionCount = 0
 
-  storage.getAllKeys((err, allKeys) => {
-    if (err) {
+  storage.getAllKeys()
+    .catch(err => {
       if (process.env.NODE_ENV !== 'production') console.warn('redux-persist/getStoredState: Error in storage.getAllKeys')
       complete(err)
-      return
-    }
-
-    let persistKeys = allKeys.filter((key) => key.indexOf(keyPrefix) === 0).map((key) => key.slice(keyPrefix.length))
-    let keysToRestore = persistKeys.filter(passWhitelistBlacklist)
+    })
+    .then(allKeys => {
```

Comment:
Hmm, I don't think this translation is quite right.

The thing is that chaining `.then` after `.catch` doesn't do what you want -- the `catch` returns a new `Promise`, one that if the `.catch` callback gets invoked, gets *resolved* with whatever that callback returned.

Quick demo:
```
$ node
> (async function () { throw new Error(); })().catch(e => {}).then(result => console.log(`got result: ${result}`))
Promise { <pending> }
> got result: undefined
```

I think you could get an accurate result by saving the `storage.getAllKeys()` promise as a local, and invoking both `.catch` and `.then` on that same promise.

... Oh hey, or alternatively: `then` [takes two arguments](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/then), one for success and then one for failure, so you could make one `.then` call and pass both. That's pretty ugly, though, in that it separates the error handler way down away from the error source, after all the code for success. Good to know for cases when the success handler is much shorter, though.

In any case I think this is a nice demonstration of why working with `.catch` and `.then` is a pain to do correctly, and `await` is much better :stuck_out_tongue: . Converting this code to `await` might be a bigger refactor, though.

Reply:
Thanks for catching this! I could do an async IIFE with try/catch; what do you think?

Reply:
I see you've done that in the new revision. It looks like a nice clean translation!
```diff
-  storage.getAllKeys((err, allKeys) => {
+  (async () => {
+    let err
+    let allKeys
+    try {
+      allKeys = await storage.getAllKeys()
+    } catch (e) {
+      err = e
+    }
```